### DISCUSSION TITLE
Why assets loads from "https://assets.tina.io/" and not from baseUrl

### DISCUSSION THREAD

**Discussion by clementroche** [tinacms/tinacms#4035] [General]
Hi, I'm auditing new potential CMS and i found that tinaCMS has huge benefits. However I noticed that assets are coming from "https://assets.tina.io/", i assume it's kinda like a s3 bucket or something on tina side. Does it means there is a bandwidth limit ? Since the asset is also on the github repository, can i have just load the asset from baseUrl instead ? (eg: https://mywebsite/uploads/image.png instead of https://assets.tina.io/image.png).

I was expecting a git-based media management too

https://tina-cloud-starter-phi-olive.vercel.app/uploads/unsplash-75EFpyXu3Wg.jpg
https://assets.tina.io/1f682f66-3301-4cf3-afe7-b38c5540041c/unsplash-75EFpyXu3Wg.jpg

**Comment by jeffsee55** [comment]
The main reason for this is because, as an editor if you upload an image to your site at `https://mywebsite/uploads/image.png` via TinaCMS, that image doesn't actually make it's way to your host. It only goes into Github, so `uploads/image.png` isn't on your host until you do a redeploy. What we used to do was temporarily serve the image from Github (eg. `https://github.com/myorg/myrepo/blob/main/public/uploads/image.png`) but if the repo was private, that image needed a short-lived token and was a big pain. 

So instead, Tina's media support aims to basically mirror the media in your repo while giving it a consistent, and immediately available place to be served from. It's something we're still collecting feedback on. There's no bandwidth limit currently with Tina's asset CDN

**Reply by clementroche** [reply]
Thank you for your reply, I understand. Would it be possible to use baseUrl in production and assets.tina.io only during preview/edit context ? The main reason is I don't want to be dependant from another cdn than my current Vercel one, that's one a the main drawbacks on other CMS i've tried.

**Reply by jeffsee55** [reply]
It's possible in a couple of ways, neither a very ergonomic unfortunately but we expect them to improve in the future. Is your production site static NextJS site?

If so, you can build your Next app during CI with the local TinaCMS server and then the CMS in edit mode would still use Tina Cloud. This only works if your 100% static though, your build command would be:

```
tinacms build --local -c "next build"
```

**Reply by clementroche** [reply]
Can't you use blob url to display assets since they're uploaded from drag n drop  ?

**Reply by jeffsee55** [reply]
Do you mean a blob stored in the editor's browser? If I'm understand you correctly, we toyed with that in the past but one big issue is that other editors wouldn't see that. Another issue is that we'd also not know when to switch back to the real URL.

**Reply by friebe** [reply]
Hi @jeffsee55 I am struggling with the same issue similarly and want to pull images from local repo instead of Tina Cloud Assets because I wanted to use the benefits of the astro Image optimization. As you noted in your comment above, using "tinacms build --local -c "xxx build"" has been very helpful for me. But as you also noted "This only works if you are running 100% static" And here comes the question, is there a good solution by now ?

If you have the time and inclination, I would be very grateful if you could briefly read my post about my exact problem here on dev.to https://dev.to/friebe/using-astro-image-optimization-benefits-with-tina-cms-cloud-in-production-build-5a01. But if you have other things on your mind and can't get to it, that's fine too :)

**Comment by joeyfigaro** [comment]
Any movement here recently?

**Reply by kulesy** [reply]
Yes, there has been movement. The piece @friebe was missing in 2024 shipped in April 2026: `tinacms build --content=local`.

**What it does**

It sources content from your local filesystem at build time (so zero Content API calls), while the generated client still points at TinaCloud at runtime. That is the split @clementroche originally asked for: your own origin in production, `assets.tina.io` in the editor.

```bash
tinacms build --content=local -c "next build"
